RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,593 @@
+      g.sy[1] = 0.0021;
+      g.sy[2] = 0.0085;
+   }
+   */
+
+    for (int i = 0; i < 3; ++i) {
+      g.sx[i] = std::sqrt(g.sx[i] * g.sx[i] + lape.xx());
+      g.sy[i] = std::sqrt(g.sy[i] * g.sy[i] + lape.yy());
+    }
+  }
+
+  // compute ladder baricenter (only in global z) for the barrel
+  auto& aveGeom = m_averageGeometry;
+  int il = 0;
+  for (int im = 0, nm = phase1PixelTopology::numberOfModulesInBarrel; im < nm; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    il = im / 8;
+    assert(il < int(phase1PixelTopology::numberOfLaddersInBarrel));
+    auto z = g.frame.z();
+    aveGeom.ladderZ[il] += 0.125f * z;
+    aveGeom.ladderMinZ[il] = std::min(aveGeom.ladderMinZ[il], z);
+    aveGeom.ladderMaxZ[il] = std::max(aveGeom.ladderMaxZ[il], z);
+    aveGeom.ladderX[il] += 0.125f * g.frame.x();
+    aveGeom.ladderY[il] += 0.125f * g.frame.y();
+    aveGeom.ladderR[il] += 0.125 * sqrt(g.frame.x() * g.frame.x() + g.frame.y() * g.frame.y());
+  }
+  assert(il + 1 == int(phase1PixelTopology::numberOfLaddersInBarrel));
+  // add half_module and tollerance
+  constexpr float module_length = 6.7f;
+  constexpr float module_tolerance = 0.2f;
+  for (int il = 0, nl = phase1PixelTopology::numberOfLaddersInBarrel; il < nl; ++il) {
+    aveGeom.ladderMinZ[il] -= (0.5f * module_length - module_tolerance);
+    aveGeom.ladderMaxZ[il] += (0.5f * module_length - module_tolerance);
+  }
+
+  // compute "max z" for first layer in endcap (should we restrict to the outermost ring?)
+  for (auto im = phase1PixelTopology::layerStart[4]; im < phase1PixelTopology::layerStart[5]; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    aveGeom.endCapZ[0] = std::max(aveGeom.endCapZ[0], g.frame.z());
+  }
+  for (auto im = phase1PixelTopology::layerStart[7]; im < phase1PixelTopology::layerStart[8]; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    aveGeom.endCapZ[1] = std::min(aveGeom.endCapZ[1], g.frame.z());
+  }
+  // correct for outer ring being closer
+  aveGeom.endCapZ[0] -= 1.5f;
+  aveGeom.endCapZ[1] += 1.5f;
+
+  /*
+  for (int jl=0, nl=phase1PixelTopology::numberOfLaddersInBarrel; jl<nl; ++jl) {
+    std::cout << jl<<':'<<aveGeom.ladderR[jl] << '/'<< std::sqrt(aveGeom.ladderX[jl]*aveGeom.ladderX[jl]+aveGeom.ladderY[jl]*aveGeom.ladderY[jl]) 
+                   <<','<<aveGeom.ladderZ[jl]<<','<<aveGeom.ladderMinZ[jl]<<','<<aveGeom.ladderMaxZ[jl]<< ' ';
+  } std::cout<< std::endl;
+  std::cout << aveGeom.endCapZ[0] << ' ' << aveGeom.endCapZ[1] << std::endl;
+  */
+
+  // fill Layer and ladders geometry
+  memcpy(m_layerGeometry.layerStart, phase1PixelTopology::layerStart, sizeof(phase1PixelTopology::layerStart));
+  memcpy(m_layerGeometry.layer, phase1PixelTopology::layer.data(), phase1PixelTopology::layer.size());
+}
+
+PixelCPEFast::GPUData::~GPUData() {
+  if (d_paramsOnGPU != nullptr) {
+    cudaFree((void*)h_paramsOnGPU.m_commonParams);
+    cudaFree((void*)h_paramsOnGPU.m_detParams);
+    cudaFree((void*)h_paramsOnGPU.m_averageGeometry);
```

Comment:
do we miss
```c++
    cudaFree((void*)h_paramsOnGPU.m_layerGeometry);
```
?

Reply:
given that is allocated together with the others:
```
cudaCheck(cudaMalloc((void**)&data.h_paramsOnGPU.m_layerGeometry, sizeof(pixelCPEforGPU::LayerGeometry)));
```
I think we miss it!

Reply:
> I think we miss it!

I agree, good catch!